barretenberg/cpp/src/barretenberg/vm2/simulation/keccakf1600.cpp

```diff
@@ -0,0 +1,98 @@
+#include "barretenberg/vm2/simulation/keccakf1600.hpp"
+
+#include <array>
+#include <cassert>
+#include <cstdint>
+
+#include "barretenberg/vm2/common/memory_types.hpp"
+
+namespace bb::avm2::simulation {
+
+namespace {
+
+MemoryValue rotate_left(MemoryValue x, uint8_t len)
+{
+    const auto x_uint64_t = x.as<uint64_t>();
+    assert(len < 64);
+    const auto out_uint64_t = (x_uint64_t << len) | x_uint64_t >> (64 - len);
+    return MemoryValue::from(out_uint64_t);
+}
+
+} // namespace
+
+// TODO: For fast simulation, we might directly call ethash_keccakf1600 from barretenberg, instead of
+// the following with no event emission. In this case, we will probably need two KeccakF1600 classes.
+KeccakF1600State KeccakF1600::permutation(const KeccakF1600State& input)
+{
+    // A lambda which transforms any one dimensional arrays of MemoryValue into a one dimensional array of uint64_t
+    auto array_to_uint64 = []<size_t N>(const std::array<MemoryValue, N>& input) {
+        std::array<uint64_t, N> output;
+        for (size_t i = 0; i < N; i++) {
+            output[i] = input[i].template as<uint64_t>();
+        }
+        return output;
+    };
+
+    // A lambda which transforms any two dimensional arrays of MemoryValue into a two dimensional array of uint64_t
+    auto two_dim_array_to_uint64 = []<size_t N, size_t M>(const std::array<std::array<MemoryValue, M>, N>& input) {
+        std::array<std::array<uint64_t, M>, N> output;
+        for (size_t i = 0; i < N; i++) {
+            for (size_t j = 0; j < M; j++) {
+                output[i][j] = input[i][j].template as<uint64_t>();
+            }
+        }
+        return output;
+    };
+
+    std::array<std::array<MemoryValue, 5>, 5> state_input_values;
```

Comment:
Maybe add a comment on why you are using MemoryValue as your main type. That is, because the bitwise gadget takes that as input. A few comments though:
* You might want to use TaggedValue. I haven't made up my mind completely, but I feel using MemoryValue gives the impression that memory is involved which it isn't.
* For readability, you might want to consider just converting when calling bitwise, and promptly converting the result to uint64. However, IDK if this will read a lot better, and it will likely be slower. So, just a consideration.